### PR TITLE
Allow hidden lifetimes in `impl Trait`

### PR DESCRIPTION
**Pull request by cramertj**
This is a revert of fc3c90cf8, which made it an error
to have a hidden lifetime inside an `impl Trait` return
type even when the hidden lifetime outlived other lifetimes
mentioned by name in the `impl Trait` bound.

Partial revert of https://github.com/rust-lang/rust/pull/49041.
cc @alexreg 

r? @nikomatsakis

### CODE REVIEW

**Comment by alexreg** [comment]
Looks good. Should we include a comment with some theoretical justification for this behaviour, as per @nikomatsakis's explanation on Zulip (implicit existential quantification over hidden lifetimes, etc.)?

**Review by Centril — APPROVED** [review]

**Comment by Centril** [comment]
@cramertj @nikomatsakis This seems right; but do we want to discuss this on a meeting / fcp-merge since it affects user observable type system behavior? Should also document this in the reference or some place less ephemeral than the code?

**Comment by matthewjasper** [comment]
This isn't sound:

```rust
trait Swap: Sized {
    fn swap(self, other: Self);
}

impl<T> Swap for &mut T {
    fn swap(self, other: Self) {
        std::mem::swap(self, other);
    }
}

fn hide<'a, 'b, T: 'static>(x: &'a mut &'b T) -> impl Swap + 'a {
    x
}

fn dangle() -> &'static [i32; 3] {
    let mut res = &[4, 5, 6];
    let x = [1, 2, 3];
    hide(&mut res).swap(hide(&mut &x));
    res
}

fn main() {
    println!("{:?}", dangle());
}
```

**Comment by alexreg** [comment]
I think we need the condition that the hidden lifetime outlives the lifetime mentioned in the opaque type?

**Comment by matthewjasper** [comment]
No, we need the currently implemented condition: `'b: 'a` in the above example.

**Comment by alexreg** [comment]
@matthewjasper Huh?

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -2011,53 +2011,6 @@ a (non-transparent) struct containing a single float, while `Grams` is a
 transparent wrapper around a float. This can make a difference for the ABI.
 "##,
 
-E0700: r##"
```
Comment (estebank) [inline src/librustc/diagnostics.rs]: I don't think we outright remove error code descriptions we don't emit anymore (at least not if they have hit stable, I'm sure).

**Comment by Dylan-DPC** [comment]
ping from triage @nikomatsakis what's the update on this?

**Comment by Dylan-DPC** [comment]
ping from triage can anyone from @rust-lang/lang review this?

**Comment by Centril** [comment]
This is blocked on fixing the soundness hole per https://github.com/rust-lang/rust/pull/57870#issuecomment-457333709.

**Comment by alexreg** [comment]
@Centril Is "blocked" the right word? There's nothing that needs to be fixed outside of this PR.

**Comment by Centril** [comment]
@alexreg Idk... I debated whether to mark it as S-waiting-on-author or S-blocked... not sure it matters; but this way we won't retriage it all the time at least.

**Comment by Dylan-DPC** [comment]
s-blocked or s-waiting-for-team should be fine at this point

**Comment by matthewjasper** [comment]
Surely this should just be closed? The soundness issue doesn't seem fixable in a way that doesn't fundamentally change this pr.

**Comment by cramertj** [comment]
@matthewjasper Yup, this is no longer just the simple revert offered by this PR given that we have to address those soundness issues. I'll close this particular PR out and we can open a new one when we figure out the correct approach.

**Pull request closed without merging**